### DISCUSSION TITLE
Strange spike in horizontal averages

### DISCUSSION THREAD

**Discussion by sam12396** [CliMA/Oceananigans.jl#1457] [Computational science!]
Hi all!
I have attached some figures of horizontally averaged fields from two recent runs. The only difference between the 2 runs is the horizontal resolution. The fine run had (100mx100mx84m) with (64,64,128) grid points. The coarse run had the same domain size and (32,32,128) points. 

In the finer resolution run, a large spike appears in the eddy viscosity, TKE, and shear. In the more coarse run, the spike does not appear, and the solution proceeds more closely to what was expected.

I have two questions for you all, 
1) What do you think would cause a spike like the one seen in the fine resolution case? Is it a physical or a numerical instability?

2) Why do you think the spike appears in the finer resolution run and not the coarse run? 

Any discussion/help is valuable and thanks for your time!
Sam

![coarse_eddy_viscosity](https://user-images.githubusercontent.com/14811146/110945210-69552a00-830b-11eb-8def-800c61844395.png)
![coarse_shear](https://user-images.githubusercontent.com/14811146/110945211-69edc080-830b-11eb-8a7e-857bc6616c5f.png)
![coarse_tke](https://user-images.githubusercontent.com/14811146/110945212-69edc080-830b-11eb-825c-ec87989089e9.png)
![fine_eddy_viscosity](https://user-images.githubusercontent.com/14811146/110945213-69edc080-830b-11eb-807a-bd4a5d5e508f.png)
![fine_shear](https://user-images.githubusercontent.com/14811146/110945214-69edc080-830b-11eb-8e06-ac00f3c00569.png)
![fine_tke](https://user-images.githubusercontent.com/14811146/110945215-6a865700-830b-11eb-996d-ead65fa9b391.png)

**Comment by francispoulin** [comment]
Thanks for sharing the results @sam12396 .  I have not used eddy viscosity here but I presume the spike reflects the fact that the velocities also develop a spike.  It could be that there is motion that could not develop in the coarse case because the dynamics could not be resolved.  Maybe the finer grid starts to resolve it.  If it were me, I would check with an even finer resolution to see if the spike persists.  If you get something similar then you have a kind of convergence, and it could be physical.  Looking at an animation of kinetic energy might be insightful as well.

**Reply by sam12396** [reply]
@francispoulin Attached below is a plot of u velocity from the fine run. There is not a 'spike' but instead, it seems to dissipate over the surface mixed layer and that reduces the shear in that same layer. 
It may also be worth noting that nothing like this spike appears in the observational data. Also, I think I am already nearing the limitations of my computing power with the current fine grid but I can give a finer resolution a shot.

 
<img width="604" alt="fine_u_vel" src="https://user-images.githubusercontent.com/14811146/110952262-01efa800-8314-11eb-80fb-cb93e48bdc55.png">

**Reply by francispoulin** [reply]
Thanks @sam12396 , I agree that there is no spike in `u`, but there is a rapid transition.  I don't know exactly what form  you are using for eddy viscosity but it pressumably depends on the gradients of the velocity.  Since `u` changes rapidly then that could explain why you have a spike in the eddy viscosity.  The next question to ask is why does `u` change rapidly and is it physical? Given that the dynamics all seems to be taking place in the top 25% of the domain, I wonder if you could narrow your depth to focus in on the interesting part of the domain.

**Reply by sam12396** [reply]
I am using Constant Smagorinsky for my closure. 

Yeah, I guess for testing I could limit the depth to a much shallower range. The full case I would like to run is 36 hours and eventually, the dynamics extend down to ~70m in our observations so we wanted to make sure to cover that range. 

I wouldn't expect a change in u that rapid based on the forcing if that is what you are asking. The forcing is a steady increase and gradual rotation of winds so I expect the velocity to look more like it does in the coarse case where it slowly increases in magnitude and deepens during the rising of the forcing.

**Reply by francispoulin** [reply]
What is the forcing exactly?  

It seems from the last figure you shared that near the right boundary, `u` is well mixed in the top 25 meters or so.  This is in contrast to the left where we have `u` very strong at the surface, and gradually decreasing in depth.  I wonder if there was a mixing event that caused the mixing near the boundary on the right, but didn't extend over the entire domain?  Looking at the evolution of `u` up to this stage might be useful to see if the behavour is reasonable or not

**Reply by sam12396** [reply]
There are wind stress and heat flux forcings at the surface which are uniform in space and vary in time. The forcing is from observations of surface meteorological fields during a typhoon. 

I am confused by your question about the mixing event on the right. For clarity's sake, the figures above are horizontally averaged fields with depth on the y-axis and time on the x-axis. There is a mixing event that should be occurring somewhat uniformly over the full horizontal extent of the domain and the forcing for that event is starting around 2 hours into the run, once the current starts to accelerate.

**Reply by francispoulin** [reply]
Ah, thanks for clarifying. That helps.  So initally, we are on the left where u = 0 throughout the depth, I presume.  Then you turn on the forcing.  After a couple of hours we see that a velocity develops near the surface.  As time passes it gets deeper and stronger.  At 9h36m we see that something has happened.  I think that this shear has become unstable and caused the fluid with non-zero velocity to mix.  This portion of well mixed fluid then persists for the duration of the simulation.  This certainly seems very plausible to me, but I have not seen the data so I don't know what you are expecting to see.  I would guess that this mixing of the surface fluid did not happen in the coarser case because the shear was not strong enough, due to the lack of spatial resolution.

**Comment by glwagner** [comment]
It looks like your boundary layer starts accelerating but does not develop turbulence until around 9 hours. The onset of turbulence is confined to a thin region at the base of the boundary layer in the coarse case, but envelops the whole boundary layer in the finer case. The development of turbulent motions is accompanied by an increase in eddy viscosity.

Here I think a primary issue is the initial laminar acceleration. Your boundary layer is unstratified, which means that the evolution of the boundary layer should probably be turbulent from t=0. I do not have much experience with `ConstantSmagorinsky`, but in my brief experience I have noticed that it tends to produce spuriously laminar flows more often than `AnisotropicMinimumDissipation`. It's easy to see why this is the case --- even unidirectional, laminar shear flow can have large eddy viscosities in `ConstantSmagorinsky`.

I think it's likely that in this simulation you would like to observe turbulent evolution from the very beginning of the simulation. At very coarse resolutions, this may require adding a lot more noise to the initial condition. I also suggest using `AnisotropicMinimumDissipation`, which is less prone to spurious laminarization and tends to produce small eddy viscosities unless the motion is truly turbulent (not just laminar shear).

The evolution of your simulation will likely be very different if it is turbulent from the beginning, since I think the "spike" is associated with a spurious transition to turbulence.

**Reply by sam12396** [reply]
This explanation makes the most physical sense to me. We do expect that the surface ocean is already somewhat turbulent, however, we don't expect the interface to be turbulent. So we should initialize with somewhat turbulent flow in the surface and wait to see the onset of turbulence at the boundary. I will do another run and include the output writers you suggested in our other conversations so that we can look at the evolution of the field.

When you say a lot more noise, what scale are you talking about? Or do I just need to play with it and find out. Also, should I only initialize with higher-order noise in the surface layer?

**Reply by glwagner** [reply]
I think it's very sensible to concentrate noise at the surface given that you would like your pycnocline to remain undisturbed at early times! You can mask the noise function with something like `exp(z / 4)` to achieve that.

Maybe bump the velocity noise amplitude to 1/10 the friction velocity (rather than 1/1000) and see what happens? You don't need to run very long to check that things make sense --- if you're unstratified, and your initial wind stress isn't too weak, you should probably observe a transition to turbulence quickly (under 15 minutes?) You'll have to evaluate whether you can tolerate that kind of initial noise for a science result, but hopefully it'll make the coarse simulations a bit more sensible. If the noise is entirely concentrated in the mixed layer, then any "memory" of it in your results will probably be wiped out shortly anyways.

**Comment by tomchor** [comment]
Hi, Sam. @glwagner and @francispoulin have said some useful things already that I won't repeat, but just a few notes from my experience with LES. This doesn't mean these things are "wrong", but they probably merit some investigation:

- When going from coarse to fine resolution you're just changing the horizontal resolution, which also changes the aspect ratio of the grid cells. Some LES models are sensitive to that, especially if the aspect ratio is larger than 3 (the coarse simulation has a ratio of ~4.8). Generally for LES you wanna keep the aspect ratio smaller than 4 (anisitropic minimum dissipation may be an exception, I'm not sure). And when doing convergence/resolution analysis you generally wanna make an effort to keep the aspect ratio constant.
- The TKE evolution of the coarse case probably indicates that your resolution is too low to support any turbulence (TKE magnitudes are pretty small). This isn't necessarily surprising, but it does mean that comparisons with turbulent simulations are challenging. 
  - One thing to check in the coarse case is the appearance of a patch of nonzero TKE starting at -60 m. If your simulation is surface-forced I don't see a reason why that patch is there.
- The fine resolution plots all look fine to me, with the exception that the eddy viscosity seems to be larger before transition to turbulence than after. That is definitely worth investigating.

Also, I should say that if your forcing isn't constant in time (which it isn't if I remember correctly) then the simulation is considerably harder to interpret. So always compare these plots to whatever is going on with your surface forcing!

**Reply by sam12396** [reply]
Your advice on the aspect ratio is a very helpful bit of guidance for this and future cases.

As for the TKE patch at ~60m in the coarse case, I do need to investigate that because it also appears in the fine case. The TKE plots for the two cases have very different colorbar limits. Because it is present in both cases, I really don't know what is going on. As you can see by the initial density profile below, the depth where this weird band starts is coincident with the beginning of the bottom water mass. Could turbulence be brought on in that lower layer just because that band is un-stratified? What is causing turbulence to start here when the turbulence in the forced surface hasn't even started?

![Screen Shot 2021-03-12 at 4 13 40 PM](https://user-images.githubusercontent.com/14811146/110999145-eb664280-834d-11eb-9397-dfe371676f3f.png)

**Reply by glwagner** [reply]
> Generally for LES you wanna keep the aspect ratio smaller than 4 (anisitropic minimum dissipation may be an exception, I'm not sure).

If I remember correctly, @johnryantaylor said he used AMD with aspect ratios of about 1/20 (but that kind of aspect ratio was only reached close to the boundary). I think AMD claims good behavior wrt grid aspect ratio, but we probably shouldn't take the claim as proven.

**Reply by tomchor** [reply]
> Could turbulence be brought on in that lower layer just because that band is un-stratified?

Not _just_ because of that, but certainly helps. Speaking _very_ generally, stratification is a deterrent of turbulence. So having no stratification make it so that any small shear that arrives there from the mixed layer can potentially cause some fluctuations. So after seeing that there's no stratification in that band (from -60 to -80) it makes a little more sense to me that there's some TKE over there after some time. The fact that it disappears in the fine simulation also supports that (it's probably still there like you said, but in much smaller magnitude than the near-surface TKE).

I'd still investigate a bit to make sure that's the case, but I'm not worried about that anymore :)

Hope this helps!

**Comment by johnryantaylor** [comment]
4:1 is what we found to be the upper limit in turbulent regions this paper:
Vreugdenhil, C., and J.R. Taylor (2018) Large-eddy simulations of stratified plane Couette flow using the anisotropic minimum-dissipation model, Phys. Fluids , 30(8), 085104.

we used a very large aspect ratio in the diffusive sublayer where the simulation transitioned from an LES to a DNS and fully resolved the scalar fluxes at the boundary (note that you just need to be a bit careful about how you define the filter width to do this).

John

**Reply by glwagner** [reply]
Very helpful, thank you! That's similar to the rules of thumb for other closures mentioned by @tomchor.
